### ISSUE TITLE
Fluid Tags no longer work for splash effects/burning

### ISSUE THREAD

**Issue by Thutmose** [MinecraftForge/MinecraftForge#9052]
**Minecraft Version:** 1.19.2

**Forge Version:** 40.1.25

**Logs:** None Relevant

**Steps to Reproduce:**
 1. Register a custom fluid, such as in the [NewFluidTest](https://github.com/MinecraftForge/MinecraftForge/blob/1.19.x/src/test/java/net/minecraftforge/debug/fluid/NewFluidTest.java)
 2. Add your new fluid to the `#minecraft:water` fluids tag (ie in /data/minecraft/tags/fluids/water.json)
 3. Jump in your new fluid, observe lack of splash
 4. Add your new fluid to the `#minecraft:lava` fluids tag (ie in /data/minecraft/tags/fluids/lava.json)
 5. Toss an item into your new fluid, observe lack of burning

**Description of issue:**

With the new fluid changes in 1.19.2, it is no longer possible to properly register a fluid to act like water and lava for the basic things it does in vanilla, such as splash sounds and burning effects. Before, this simply consisted of adding the fluid to the appropriate tag, and then it would work. Now, the tags do nothing for this, as everything was replaced by references directly to `ForgeMod.WATER_TYPE` and `ForgeMod.LAVA_TYPE`

Example of such changes in the patch to Entity.java:

[[517/518] prevents checking depth from working](https://github.com/MinecraftForge/MinecraftForge/blob/1.19.x/patches/minecraft/net/minecraft/world/entity/Entity.java.patch#L517-L518)
[[424/425] stops the splash effects from working](https://github.com/MinecraftForge/MinecraftForge/blob/1.19.x/patches/minecraft/net/minecraft/world/entity/Entity.java.patch#L424-L425)
[[232/233] Prevents water on eyes tag test from working](https://github.com/MinecraftForge/MinecraftForge/blob/1.19.x/patches/minecraft/net/minecraft/world/entity/Entity.java.patch#L232-L233)

In general, all of the places where the tag check was placed with a direct reference to a fluid type prevents the use of the tags.

**Comment by KnightMiner** [comment]
Adding them to the tags to get vanilla behavior is not desired, it pollutes those tags that are expected to be just vanilla fluids. Have you tried implementing those behaviors using the new fluid system?

**Comment by Thutmose** [comment]
Yes, the issue is in deciding on the optimal way to implement those behaviours.

For the splash effect for example, vanilla does it as follows:
```java
      if (this.getVehicle() instanceof Boat) {
         this.wasTouchingWater = false;
      } else if (this.updateFluidHeightAndDoFluidPushing(FluidTags.WATER, 0.014D)) {
         if (!this.wasTouchingWater && !this.firstTick) {
            this.doWaterSplashEffect();
         }

         this.resetFallDistance();
         this.wasTouchingWater = true;
         this.clearFire();
      } else {
         this.wasTouchingWater = false;
      }
 ```
 this is the same in 1.18.2 and 1.19.2, what changed was the implementation of `updateFluidHeightAndDoFluidPushing`, where forge replaces the actual tag check.

With the existing fluid system, I can run things while the entity is in the fluid, ie via `canExtinguish` or any of the other similar methods, but I do not have an event/test for when the entity leaves the fluid, except by say checking it every tick.

If the expected solution is to check for every tick, this doesn't work for non-living entities, as there isn't a general update event for them.

**Comment by ChampionAsh5357** [comment]
The correct way to do this would be to add two methods that get called whenever a fluid is entered and exited. In those cases, you would need to keep track of the previous state of the fluid each tick. Then, if somebody wanted to add logic for splashing, it would be handled there.

Additionally, the sounds should be relegated to the sound action system.

**Comment by ChampionAsh5357** [comment]
Any support for the tag system was vetoed when this was being designed two years ago, so we won't be adding such a thing back for custom fluids.

**Comment by Thutmose** [comment]
enter/exit methods would be sufficient, along with I guess the mod adding some way to track the previous state of the mob. for the sounds, that can be done by using an AT to make `entity.doWaterSplashEffect` public, then just calling that during the on enter.

Why was the tag support decided against? can you link the relevant discussion about that so I can read about it? I find it somewhat strange, as vanilla still uses the water tag for quite a few things, which I see are all still checked, but often then followed by a re-direct to the FluidType checks.

**Comment by ChampionAsh5357** [comment]
> can you link the relevant discussion about that so I can read about it?

The discussion is in a private channel, so I cannot. Essentially though, it boils down to two points: Forge doesn't use logic tags because of synchronization whether between Forge Client/Server or Vanilla Client/Server, and the most of the people who contributed to the initial stages of the PR believe that tags should not be used for world logic.

**Comment by Thutmose** [comment]
It is fine, it was explained in the discord channel, and I agree with the reasons given there. I generally am curious as to why things are done in certain ways, hence asking for links if they were present.
